### ISSUE TITLE
Loss_adv_p increasing, while loss_D decreasing. Is it correct?

### ISSUE THREAD

**Issue by John1231983** [hfslyc/AdvSemiSeg#7]
Hello, I want to ask one question related to the `loss_adv_p` and `loss_D`. As the log, the `loss_adv_p` tends to increase while the `loss_D` tends to decrease. We expect that `loss_D` will decrease, it is correct; however, how about the `loss_adv_p`? I found one interesting thing that is `args.lambda_adv_pred`, you can set it to `0.5`, then the `loss_adv_p` will decrease but the loss `loss_D` will increase. They looks opposite side  Thanks
```
iter =        0/   20000, loss_seg = 3.045, loss_adv_p = 0.695, loss_D = 0.693, loss_semi = 0.000
```
and at 180 iters
```
iter =      180/   20000, loss_seg = 0.490, loss_adv_p = 1.358, loss_D = 0.572, loss_semi = 0.000
```

**Comment by hfslyc** [comment]
loss_adv_p is essentially the inverse loss of loss_D. We expect them to achieve a good balance during the adversarial training. Please refer to equation (1) and (4) in our paper. We use lambda_adv to control the balance point between loss_adv_p and loss_D. Does that answer your question?

**Comment by John1231983** [comment]
Great. You have mentioned that your task tries  to minimize loss_D and maximize loss_adv_p. However, when we maximize loss_adv_p, the `  loss_seg_total = loss_seg+ lambdax loss_adv_p ` in ep (2) tends to increase. But we want to minimize the eq (2). Does it make sense?

One more thing, your implementation use 
You used
```
for param in model_D.parameters(): 
     param.requires_grad = False 
loss = loss_seg + args.lambda_adv_pred * loss_adv_pred 
loss.backward() 
```
Does it same as the below?
```
loss = loss_seg + args.lambda_adv_pred * loss_adv_pred 
loss_seg.backward() 
```
By delete the first two lines and change loss.backward() to loss_seg.backward()because loss_adv_pred does not do backward() so D not trained.

**Comment by hfslyc** [comment]
We are trying to minimize loss_D with respect to the discriminator and
minimize loss_adv with respect to the segmentation network.

**Comment by John1231983** [comment]
Thanks. I have updated something in the above  question related to implementation. Could you also look at it? Thanks for your answer

**Comment by hfslyc** [comment]
No. It's not equivalent. By doing that you are not performing the adversarial training since you are not minimizing loss_adv.

**Comment by John1231983** [comment]
@hfslyc : Thanks. I have one more question. In my case, the both `loss_adv_p` and `loss_D` are decreasing. What does happen in my case? Do we expect that `loss_adv_p` increases and `loss_D` decrease?
